pubsub/google/cloud/pubsub_v1/publisher/futures.py

```diff
@@ -0,0 +1,179 @@
+
+    def cancel(self):
+        """Publishes in Pub/Sub currently may not be canceled.
+
+        This method always returns False.
+        """
+        return False
+
+    def cancelled(self):
+        """Publishes in Pub/Sub currently may not be canceled.
+
+        This method always returns False.
+        """
+        return False
+
+    def running(self):
+        """Publishes in Pub/Sub currently may not be canceled.
+
+        This method always returns True.
+        """
+        return True
+
+    def done(self):
+        """Return True if the publish has completed, False otherwise.
+
+        This still returns True in failure cases; checking :meth:`result` or
+        :meth:`exception` is the canonical way to assess success or failure.
+        """
+        return self._exception is not None or self._result is not None
+
+    def result(self, timeout=None):
+        """Return the message ID, or raise an exception.
+
+        This blocks until the message has successfully been published, and
+        returns the message ID.
+
+        Args:
+            timeout (Union[int, float]): The number of seconds before this call
+                times out and raises TimeoutError.
+
+        Returns:
+            str: The message ID.
+
+        Raises:
+            ~.pubsub_v1.TimeoutError: If the request times out.
+            Exception: For undefined exceptions in the underlying
+                call execution.
+        """
+        # Attempt to get the exception if there is one.
+        # If there is not one, then we know everything worked, and we can
+        # return an appropriate value.
+        err = self.exception(timeout=timeout)
+        if err is None:
+            return self._result
+        raise err
+
+    def exception(self, timeout=None, _wait=1):
+        """Return the exception raised by the call, if any.
+
+        This blocks until the message has successfully been published, and
+        returns the exception. If the call succeeded, return None.
+
+        Args:
+            timeout (Union[int, float]): The number of seconds before this call
+                times out and raises TimeoutError.
+
+        Raises:
+            TimeoutError: If the request times out.
+
+        Returns:
+            Exception: The exception raised by the call, if any.
+        """
+        # If no timeout was specified, use inf.
+        if timeout is None:
+            timeout = float('inf')
+
+        # If the batch completed successfully, this should return None.
+        if self._result is not None:
+            return None
+
+        # If this batch had an error, this should return it.
+        if self._exception is not None:
+            return self._exception
+
+        # If the timeout has been exceeded, raise TimeoutError.
+        if timeout <= 0:
+            raise exceptions.TimeoutError('Timed out waiting for result.')
+
+        # Wait a little while and try again.
+        time.sleep(_wait)
+        return self.exception(
```

Comment:
this is likely to hit the recursion limit. Just use a `threading.Event` and wait on it here and set it in `_trigger`.